### ISSUE TITLE
Python fails silently in an apply if a resource type is mistyped

### ISSUE THREAD

**Issue by jkodroff** [pulumi/pulumi#10637]
From my debugging session with @AaronFriel, this code fails silently - the loop only executes once. (The resource created should be `aws.ec2.Route` not `aws.ec2.route`):

```python
    route_table_ids = vpc.route_tables.apply(lambda route_tables: map(
        lambda route_table: route_table.id, route_tables))

    for route_table_id in route_table_ids:
        aws.ec2.route(
            f"route-to-cloudwan-subnet{route_table_id}",
            aws.ec2.RouteArgs(
                core_network_arn=args.core_network_arn,
                destination_cidr_block="0.0.0.0/0",
                route_table_id=route_table_id,
            ),
            opts=pulumi.ResourceOptions(
                provider=provider,
                depends_on=vpc_attachment,
            )
        )
```

Josh owes the Platform Team minimal steps to repo but is working on workshop code on a deadline as he files this. Wanted to make sure I got this at least filed.

**Comment by iwahbe** [comment]
Could you provide a full repro? I'm having trouble figuring out what type `aws.ec2.route` is.

**Comment by jkodroff** [comment]
@iwahbe This isn't minimal, but I believe this program demonstrates the same behavior:

```python
"""A Python Pulumi program"""

import pulumi
import pulumi_aws as aws
import pulumi_awsx as awsx
from typing import Sequence

region = aws.get_region()
project = pulumi.get_project()

# TODO:
# - Tags

hub_and_spoke_supernet = pulumi.Config().require("hub-and-spoke-supernet")

tgw = aws.ec2transitgateway.TransitGateway(
    "tgw",
    aws.ec2transitgateway.TransitGatewayArgs(
        description=f"Transit Gateway - {project}",
        default_route_table_association="disable",
        default_route_table_propagation="disable"
    )
)

aws.ec2transitgateway.RouteTable(
    "spoke-vpc-route-table",
    aws.ec2transitgateway.RouteTableArgs(
        transit_gateway_id=tgw.id,
        tags={
            "Name": "Spoke Route Table",
        }
    )
)

aws.ec2transitgateway.RouteTable(
    "post-inspection-route-table",
    aws.ec2transitgateway.RouteTableArgs(
        transit_gateway_id=tgw.id,
        tags={
            "Name": "Post-Inspection Route Table",
        }
    )
)

inspection_vpc = awsx.ec2.Vpc(
    "inspection-vpc",
    awsx.ec2.VpcArgs(
        cidr_block="10.129.0.0/24",
        availability_zone_names=[
            pulumi.Output.concat(region.name, "a"),
            pulumi.Output.concat(region.name, "b"),
        ],
        subnet_specs=[
            awsx.ec2.SubnetSpecArgs(
                type=awsx.ec2.SubnetType.PUBLIC,
                cidr_mask=28,
            ),
            awsx.ec2.SubnetSpecArgs(
                type=awsx.ec2.SubnetType.PRIVATE,
                cidr_mask=28,
                name="inspection",
            ),
            awsx.ec2.SubnetSpecArgs(
                type=awsx.ec2.SubnetType.ISOLATED,
                cidr_mask=28,
                name="tgw"
            ),
        ]
    )
)


def create_tgw_routes(subnet_ids: Sequence[str]):
    print("IN something()")
    print(f"  subnet_ids={subnet_ids}")
    for subnet_id in subnet_ids:
        print("FOR LOOP")
        route_table = aws.ec2.get_route_table(
            subnet_id=subnet_id
        )

        # Fails silently due to missing name:
        aws.ec2.Route(
            route_table_id=route_table.id,
            destination_cidr_block=hub_and_spoke_supernet,
            transit_gateway_id=tgw.id,
        )

        # Works:
        aws.ec2.Route(
            f"tgw-route-{subnet_id}",
            aws.ec2.RouteArgs(
                route_table_id=route_table.id,
                destination_cidr_block=hub_and_spoke_supernet,
                transit_gateway_id=tgw.id,
            ),
        )


inspection_vpc.private_subnet_ids.apply(
    lambda subnet_ids: create_tgw_routes(subnet_ids))

pulumi.export("private_subnet_ids", inspection_vpc.private_subnet_ids)
```

**Comment by iwahbe** [comment]
This example can be reduced to 
```python
import pulumi
import pulumi_aws as aws
import pulumi_awsx as awsx
from typing import Sequence

region = aws.get_region()

inspection_vpc = awsx.ec2.Vpc(
    "inspection-vpc",
    awsx.ec2.VpcArgs(
        cidr_block="10.129.0.0/24",
        availability_zone_names=[
            pulumi.Output.concat(region.name, "a"),
            pulumi.Output.concat(region.name, "b"),
        ],
        subnet_specs=[
            awsx.ec2.SubnetSpecArgs(
                type=awsx.ec2.SubnetType.PUBLIC,
                cidr_mask=28,
            ),
            awsx.ec2.SubnetSpecArgs(
                type=awsx.ec2.SubnetType.PRIVATE,
                cidr_mask=28,
                name="inspection",
            ),
            awsx.ec2.SubnetSpecArgs(
                type=awsx.ec2.SubnetType.ISOLATED, cidr_mask=28, name="tgw"
            ),
        ],
    ),
)


def create_tgw_routes(subnet_ids: Sequence[str]):
    print("SHOULD BREAK HERE")
    raise Exception("🔥")


inspection_vpc.private_subnet_ids.apply(create_tgw_routes)
```

We will see `SHOULD BREAK HERE` but not the `🔥`.

**Comment by jkodroff** [comment]
@iwahbe Probably doesn't help too much, but I suspect you can remove all the `VpcArgs` as well.

**Comment by jkodroff** [comment]
For anyone else who might be hitting this issue and needs a workaround, you can get some debugging information by temporarily surrounding your code with a `try/except`:

```python
try:
  # create some resources
except Exception as e:
  print(e)
```

**Comment by jkodroff** [comment]
Per discussion with @AaronFriel, this issue can lead to unexpected deletion of resources if resources are defined after the silently caught exception. He has suggested that this may need to be escalated to a p1.

**Comment by Frassle** [comment]
@jkodroff I'm also struggling to repro this, @iwahbe's example above seems to work fine for me (the program fails because the apply throws an exception).

Wondering if I've misunderstood what this issue is actually about? It looks like there's maybe three different things here:
1. Giving the wrong type for a resource (`route` instead of `Route`)
2. Not giving a resource a name (no `name=` parameter)
3. Exceptions not showing in apply (Ian's example above)

Can I confirm what exactly you had issues with as all three of the above seem to work to me.

**Comment by jkodroff** [comment]
@Frassle The issue is that exceptions are swallowed when they occur within Apply. The use of `route` versus `Route` or the lack of a `name=` were just things that happened to trigger an exception in my case.

If you see a proper stack trace from @iwahbe's example, then the bug is not exhibiting for you and it makes me wonder if it's specific to macOS (or at least not-Windows).

**Comment by Frassle** [comment]
So I tracked this down to the runtime not waiting for all applys to run before quitting.
The other language SDKs have global registration lists for Outputs so we always await them before program close, but we're missing that logic in python.

The simplest workaround is to just return a dummy value (like `None`) from the apply and `pulumi.export` it so that it's definitely waited for. To fix I'll have a look into tracking all outputs in Python like we do for other languages.

**Comment by julienp** [comment]
This was fixed in 3.112 with https://github.com/pulumi/pulumi/pull/15744 and https://github.com/pulumi/pulumi/pull/16057
